### ISSUE TITLE
Discussion: Re-design for Plugins

### ISSUE THREAD

**Issue by orta** [graphql/graphiql#978]
👋  - Here's a general direction for where we can take the GraphiQL. It takes design cues from Storybook JS and VSCode. It aims to have quite a few UI chrome entry points for plugins to be able to work from. This ranges from adding a sidebar button which can either control a sidebar or the whole UI, to toolbar items and tabs on request/reponse.

![GraphiQL](https://user-images.githubusercontent.com/49038/66406880-346bd780-e9a1-11e9-82c9-669a21436eb3.png)

This is roughly how I see each section:

![Area overview](https://user-images.githubusercontent.com/49038/66407623-5d409c80-e9a2-11e9-84cb-30ffa2081b7d.png)

# Notes

- This top bar (current query description) is questioned every time I show the design. Its goal is to provide scanability and to give an official "title" to every query. The title is basically either the name of the query + up to 3 vars. If there's no name, we try infer it from the root query/mutation fields. This has 2 main advantages: when you have many tabs with you can know what the page is, when we show summaries of tabs there is a consistent naming.

- I've built out a consistent way to show the different types in GraphQL:  <img width="217" alt="Screen Shot 2019-10-08 at 9 05 15 AM" src="https://user-images.githubusercontent.com/49038/66412639-c9bf9980-e9aa-11e9-96c5-4e5fc2b5536c.png">

- I've tried to always use one highlight colour which is used in different shades: which means that can be customized by a consumer. Artsy might have used the purple from it's logo. The default value is the turquoise colour from the GrpahQL logo.

- Docs currently lives in the sidebar, as a sub navigation which is pushed on top of whatever is  on the left, in a similar stack metaphor to how iOS does navigation. There's a back button and that pops the stack. I'm open to this changing, it can cover useful UI by doing this. The trade-off is that there needs to be a special section just for introspection of the code

![Search Result (Docs)](https://user-images.githubusercontent.com/49038/66412259-0939b600-e9aa-11e9-90ee-4951ef189aa7.png)

- Things missing from design spec:
  - Run button
  - Sure button
  - Error tab? 
  - Explorer style query generation 

- Things I'm not sold on and can change:
  - The sidebar reference to a query, feels like it needs to be tighter somehow

# Example plugins

Ideally most of the chrome in GraphiQL comes from the plugin infra, that means we can be sure to build a scalable plugin API which works with massive schemas.

### Schema

The default sidebar tab: provides an entry point into your schema and allows for deeper introspection of your current query. It probably doesn't need to interact with requests or responses. It provides a way to dig into the documentation of the schema.

![image](https://user-images.githubusercontent.com/49038/66410580-3042b880-e9a7-11e9-9509-1fc64c751fb3.png)

### History

It would listen to every request made, check to see if it was different from the last one and keep track of the diffs. It would add a toolbar item, which when clicked showed a list of the requests and you can click to set the query string to be one you hit in the past

![History](https://user-images.githubusercontent.com/49038/66408077-228b3400-e9a3-11e9-8878-775f5de6bbd2.png)

### Stored Queries

It can add a tab to the request toolbar to save queries, as well as having a consistent way to access queries at runtime. 

For example Artsy keeps all its stored queries in a [JSON file](https://github.com/artsy/metaphysics/blob/master/src/data/complete.queryMap.json), so the configuration for this plugin could have a function to index or search these queries. The same could be done against the Apollo server [automatic persisted queries](https://blog.apollographql.com/automatic-persisted-queries-and-cdn-caching-with-apollo-server-2-0-bf42b3a313de)

![History Copy](https://user-images.githubusercontent.com/49038/66409187-0be5dc80-e9a5-11e9-85dd-18640dc33ceb.png)

### Playground

Provides a toolbar icon to the requests section of graphiql which shows a tab which  lets you eval JavaScript against the response. Extra kudos points if that has type completion for that JSON ;)

![image](https://user-images.githubusercontent.com/49038/66411306-63d21280-e9a8-11e9-838f-6dcd1e5be2a8.png)

---

[Sketch file](https://ortastuff.s3.us-east-1.amazonaws.com/graphiql/graphiql.sketch)

**Comment by domyen** [comment]
Awesome work @orta! I support the general direction so will limit my feedback to design for now:

**Icons** 
Icons for technical products can sometimes be tough to understand because they boil down an abstraction to a 14x14 symbol. I wonder if a simpler shape with the first letter inside might be simpler to comprehend at a glance.

**Layout**
I think there's a visual disconnect between the query panel and the response panel. If they were conjoined (e.g., remove the margin between them) it might make the relationship between them clearer. Same for the query title header and the query/response panel. 

**Run button** It's pretty useful for new users that might be unfamiliar with the UI. I used it in the past and still do.

**Comment by orta** [comment]
I love this feedback ❤️ - will give that a look

**Comment by acao** [comment]
I love the idea of adding tabs to each of the three default panes. we will need to make it so that the default tab is overrideable. do you remember my idea of a diagnostics/console area that slides up from the bottom? similar to vscode or codesandbox? gives a nice birds eye view when you're linting large queries, etc. might be nice for things like response headers, though we can just make that a secondary tab on the right pane.

Also for that legend, i like the icons, and the look and feel is close to the official look and feel which is great. For the legend labels, we need to use spec language - for example i think primitive is scalar? 

Plugin notes: The side icon bar and panel region will itself be provided via a plugin. Also, I'd like to encapsulate the default layout in a simple layout component that doesn't need to pass any props, because some people want to override the default three pane layout, or we can just encourage them to compose the existing components however they want. We need to make this capability first order, in documentation/etc because it's where a lot of people get stuck very quickly.

**Comment by benjie** [comment]
This looks incredible, @orta; thanks for sharing your progress!

Here's a couple of questions/notes I have that I couldn't see the answers to in your designs:

One of the popular things in GraphQL Playground was the ability to browse through the docs and keep the previous layer and a bit open for context - effectively keeping a couple of the previous states of the navigation stack that you mention visible. I think we would like to enable this via a plugin; how do you see that fitting into the design?

Another thing that's commonly requested is a header and URL editor. I'm not sure if this would comfortably fit in the sidebar due to lack of width (URL bar especially). Maybe we could address this with plugin hooks to the "Current Query Description" area; I'm thinking a plugin could do something like adding an icon to the left/right that would open a header editing popover/modal. I'm not sure what this means for usage on mobile though. And perhaps that popover/modal should be baked into GraphiQL for all plugins to use rather than having each reinvent the wheel?

How do you see this mapping to mobile? Storybook is surprisingly good on mobile I discovered a few days ago, so perhaps we could take some inspiration from there.

Just want to say again - absolutely incredible work! :raised_hands:

**Comment by acao** [comment]
im thinking the query tab itself could have a network tab and thus the url, verb, headers, etc

the notes about the navigation trail are critical, agreed thanks @benjie

**Comment by orta** [comment]
There's some great explorer ideas here: https://github.com/OneGraph/graphiql-explorer/issues/10

**Comment by dwwoelfel** [comment]
> There's some great explorer ideas here: [OneGraph/graphiql-explorer#10](https://github.com/OneGraph/graphiql-explorer/issues/10)

The explorer seems like it would fit really well as a plugin in the leftmost column.

We also have a [code exporter plugin](https://github.com/OneGraph/graphiql-code-exporter) that would fit well in the rightmost column.

**Comment by acao** [comment]
Definitely something we’re planning on!

**Comment by orta** [comment]
Do peeps have opinions on these sort of type signifiers?

<img width="689" alt="Screen Shot 2019-11-23 at 1 13 27 PM" src="https://user-images.githubusercontent.com/49038/69483272-3542b680-0df3-11ea-91f7-0076cb4d9b26.png">

**Comment by acao** [comment]
> > There's some great explorer ideas here: [OneGraph/graphiql-explorer#10](https://github.com/OneGraph/graphiql-explorer/issues/10)
> 
> The explorer seems like it would fit really well as a plugin in the leftmost column. I think we've implemented most of the feedback from that issue.
> 
> We also have a [code exporter plugin](https://github.com/OneGraph/graphiql-code-exporter) that would fit well in the rightmost column.

I plan on introducing this pretty early on actually. I'd like to make the doc explorer and query histories into plugins, as ways of defining what the initial plugin interface should be. the idea is to abstract CM and monaco details enough that `addCommand()` will translate over fluidly into a transformed experience for 1.0.0

**Comment by acao** [comment]
> Do peeps have opinions on these sort of type signifiers?

I feel like the middle ones with the square border box are easier to identify than the ones on the right without a square, but there could be a11y issues I'm not aware of. Thanks for digging in on this today @orta! Happy Birthday!

**Comment by CFSJthomas** [comment]
Dark mode please.

**Comment by ttzztztz** [comment]
In my opinion, The dark mode should be supported. Latest version of Safari, Chrome, Firfox support the CSS media query below:

```css
@media (prefers-color-scheme: dark) {

}
```

By writing some CSS codes, we can make it support dark mode easily, however, design picture for dark mode is needed.

**Comment by orta** [comment]
I'm going to make dark-mode request comment as off-topic, there is plans to support theming and dark mode is effectively a theme

**Comment by acao** [comment]
nice, theme-ui has some handy stuff for this

**Comment by Chhekur** [comment]
1. We can actually provide a feature to preview the response in the form of tree like structure for better visualization.
2. We can actually provide a visual programming tool for writing queries, so that user won't actually have to remember the query attributes

**Comment by acao** [comment]
https://github.com/graphql/graphiql/blob/master/working-group/agendas/2020-03-10.md

heres the agenda for the working group tomorrow! we will be discussing this and other topics please open a pr to add yourself to the agenda!

**Comment by alinauroz** [comment]
We can add **tabs** option to Graphiql. This will help help a lot.  When we have to switch between different queries multiple times, it becomes a little complicated to use history or commenting. I did a little changes in @orta proposed UI.
![66406880-346bd780-e9a1-11e9-82c9-669a21436eb3](https://user-images.githubusercontent.com/38759611/76449321-aed7d200-63ed-11ea-983a-b5a2cf7b03c2.png)

**Comment by acao** [comment]
thanks for that! @orta has a design that includes tabs somewhere, and we are working towards tabs support

**Comment by alinauroz** [comment]
> * We can actually provide a feature to preview the response in the form of tree like structure for better visualization.
> * We can actually provide a visual programming tool for writing queries, so that user won't actually have to remember the query attributes

@Chhekur dont you think treeview structure is already implemented. Look this image
![DB25E813-47A1-4835-A59E-FCEAE6282FAF](https://user-images.githubusercontent.com/38759611/76515949-54398700-647c-11ea-8ac3-753531ed9d3d.jpeg)

But i think it would be a better approach if we move expand/collapse button with the nodes. But we have to make this feature optional for users through plugins or extensions

I have a code that implements a relatively better treeview. Im looking for a container to view it. Graphiql is using CodeMirror to show response and we cant render html in CodeMirror.
 I will make a pr soon

**Comment by acao** [comment]
graphql explorer plugin is a great “visual programming tool”. @Chhekur may be referring to a tree view using nodes/edges for json results? we will be providing plugins for handling different result formats , such as json, mdx, xml or yaml (all have been requested), and tabs for displaying reaults visually. any such proposal we can come up with sounds great. 

working on the plugin proposal template today!

**Comment by Chhekur** [comment]
> graphql explorer plugin is a great “visual programming tool”. @Chhekur may be referring to a tree view using nodes/edges for json results? we will be providing plugins for handling different result formats , such as json, mdx, xml or yaml (all have been requested), and tabs for displaying reaults visually. any such proposal we can come up with sounds great.
> 
> working on the plugin proposal template today!

Yes I was talking about representation of data in the form of node/edges, it'll be very easy to understand for user
